Put operator SSH keys to service VMs

byucesoy opened the pull request:
We deployed [this PR](https://github.com/ubicloud/ubicloud/pull/1744), but it caused failures in VM provisioning. This line was the culprit;

```
ssh_public_keys = params.fetch("ssh_public_keys", params.fetch("ssh_public_key"))
```

Here is what happens. Name of the variable sent from control plane was `ssh_public_key` (i.e. singular) but we are changing it to be `ssh_public_keys` (i.e plural). We added the second branch to fetch, where we fetch the `ssh_public_key` variable if `ssh_public_keys` is not found. In theory, we could deploy data plane that contains the above line first and install rhizome. This would ensure that no matter if the control plane code is old or new, the code would work. However, in the above line, right branch is evaluated first, without checking if `ssh_public_keys` exist or not. Then it throws error if `ssh_public_key` is not found. This makes sense in hindsight; after all variables of the function needs to be evaluated before the function. That means, in the most up-to-date control plane version of the code, where we use the variable name as `ssh_public_keys`, we always throw error due to lack of `ssh_public_key`. The fix is simple, we add a third branch:

```
ssh_public_keys = params.fetch("ssh_public_keys", params.fetch("ssh_public_key", ""))
```

This ensures that evaluation of second branch never throws error. However, in my opinion, having third branch has some problems;
- Three branches for fetching a simple variable is too much.
- We lose the ability to ever throw an error if neither `ssh_public_keys` nor `ssh_public_key` is sent.
- Branching is only useful during the deployment of this patch.

Thus, I added a final commit, that is supposed to be deployed last and remove the branching. The final version of the code looks like this;

```
ssh_public_keys = params.fetch("ssh_public_keys")
```

enescakir commented:
When you assign a default value using a block, it isn't evaluated upon calling. I learnt it from @velioglu https://github.com/ubicloud/ubicloud/pull/1614#discussion_r1608872614

```ruby
ssh_public_keys = params.fetch("ssh_public_keys") { params.fetch("ssh_public_key") }
```

fdr commented on rhizome/host/lib/vm_setup.rb:
```diff
@@ -462,15 +462,15 @@ def generate_swap_config(swap_size_bytes)
     shell: /bin/bash
     ssh_authorized_keys:
-      - #{yq(public_key)}
+#{public_keys.map { "      - #{yq(_1)}" }.join("\n")}
```
Oh man, that we're still templating YAML (a prototype carryover) is such a bummer, but not for this patch to fix.

enescakir commented on spec/prog/vm/nexus_spec.rb:
```diff
@@ -227,6 +227,8 @@ def requested_disk_size(st)
       expect(vm).to receive(:nics).and_return([nic]).at_least(:once)
       expect(vm).to receive(:cloud_hypervisor_cpu_topology).and_return(Vm::CloudHypervisorCpuTopo.new(1, 1, 1, 1))
       expect(vm).to receive(:pci_devices).and_return([pci]).at_least(:once)
+      prj.set_ff_vm_public_ssh_keys(["operator_ssh_key"])
```
The feature flag introduced in the next commit

<img width="932" alt="Screenshot 2024-07-16 at 12 11 33" src="https://github.com/user-attachments/assets/e2f3fac1-3022-4a91-8601-0c538ad8661b">

byucesoy replied:
Moved the test changes to next commit.

enescakir reviewed (approved):
You might write a brief explanation for the last commit detailing why we removed that part, for example we keep it at first place for zero downtime deployment etc. This will help clarify the reason behind the removal for anyone reviewing individual commits in the future via blame

byucesoy commented:
> You might write a brief explanation for the last commit detailing why we removed that part, for example we keep it at first place for zero downtime deployment etc. This will help clarify the reason behind the removal for anyone reviewing individual commits in the future via blame

Added a brief explanation. Running the E2E tests now. I'll deploy this after they pass.
